### PR TITLE
Refactor h2 server to allow proper flow control

### LINKED ISSUE #730: ensure end-to-end flow control
Initial tests show that linkerd does not properly apply flow control across client and server streams.

For flow control to work properly, stream transport writes _must not_ complete until the window is opened. I think that this does not currently behave correctly, so that linkerd blindly reads an entire stream even when the next hop cannot receive it, eventually causing linkerd to OOM.

### PR DESCRIPTION
**Pull request by olix0r**
##### Problem

In the initial implementation of h2, our managed multiplexing in the Netty
stream so that H2 listeners created a Transport for each client stream (and not
per-connection, as is done for other protocols).

We also learned that Netty's (experimental) Http2MultiplexCodec has
flow-control behavior that does not currently play well with linkerd's model.
Specifically:
- Write promises may be satisfied before the entire data frame is written.
- WINDOW_UPDATE messages are automatically issued by the codec as soon as a
  data frame is read. In linkerd, we want to defer this update until the data
  frame has been completely forwarded.

The combination of these issues makes it difficult if not impossible for
linkerd to manage flow control properly.

##### Solution

This change moves linkerd's H2 server from netty's Http2Codec to
Http2FrameCodec.  The FrameCodec provides only the encoding/decoding subset of
features, and is already used by our client.  The Netty4ClientStreamTransport
is generalized as the Netty4StreamTransport, which is now used by both clients
and servers.

Now, h2 listeners create a Netty4StreamTransport for each client connection.
Netty4ServerDispatcher implements multiplexing logic---similar, but inverse, to
that of Netty4ClientDispatcher.  This allows linkerd to manage flow control
explicitly and makes the Client and Server code paths more congruent.

The h2 end-to-end test has been extended to validate flow control behavior.

Fixes #730

### CODE REVIEW

File: router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    // Put two windows' worth of data on the stream
+    val release0, release1 = new Promise[Unit]
+    def releaser(p: Promise[Unit]) = () => {
+      p.setDone()
+      Future.Unit
+    }
+
+    // The first frame is too large to fit in a window. It should not
+    // be released until all of the data has been flushed.  This
+    // cannot happen until the reader reads and erleases some of the
```
Comment (adleong) [inline router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala]: minor typo

File: router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    assert(frames.map(_.buf.length).sum == read)
+    assert(read == WindowSize)
+    assert(!release0.isDefined && !release1.isDefined)
+
+    // At this point, we've read an entire window of data from
+    // the stream without releasing any of it.  Subsequent reads
+    // should not complete until data is released.
+    val rf0 = reader.read()
+    for (_ <- 0 to 10000) assert(!rf0.isDefined)
```
Comment (adleong) [inline router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala]: what's the purpose of testing this 10000 times?  is this reliable?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 3]
```diff
@@ -32,24 +38,33 @@ class Netty4ClientDispatcher(
 
   // TODO handle overflow
   private[this] val _id = new AtomicInteger(BaseStreamId)
-  private[this] def nextId() = _id.getAndAdd(2)
-  private[this] val streams = new ConcurrentHashMap[Int, Netty4ClientStreamTransport]
+  private[this] def nextId(): Int = {
+    val id = _id.getAndAdd(2)
+    if (id > MaxStreamId && closed.compareAndSet(false, true)) {
+      writer.goAwayProtocolError(Time.Top)
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: On stream id overflow, you send a GOAWAY?  Not a stream reset?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 4]
```diff
@@ -32,24 +38,33 @@ class Netty4ClientDispatcher(
 
   // TODO handle overflow
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: is this TODO complete?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 5]
```diff
@@ -32,24 +38,33 @@ class Netty4ClientDispatcher(
...
   private[this] val _id = new AtomicInteger(BaseStreamId)
-  private[this] def nextId() = _id.getAndAdd(2)
-  private[this] val streams = new ConcurrentHashMap[Int, Netty4ClientStreamTransport]
+  private[this] def nextId(): Int = {
+    val id = _id.getAndAdd(2)
+    if (id > MaxStreamId && closed.compareAndSet(false, true)) {
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: what happens if the stream id overflows but the CAS fails?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 6]
```diff
@@ -6,22 +6,28 @@ import com.twitter.finagle.stats.{StatsReceiver, NullStatsReceiver}
...
 import java.util.concurrent.ConcurrentHashMap
-import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+import scala.collection.JavaConverters._
+import scala.util.control.NoStackTrace
 
 object Netty4ClientDispatcher {
   private val log = Logger.get(getClass.getName)
   private val BaseStreamId = 3 // ID=1 is reserved for HTTP/1 upgrade
   private val MaxStreamId = math.pow(2, 31) - 1
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: is it intentional that this is a double?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 7]
```diff
@@ -32,24 +38,33 @@ class Netty4ClientDispatcher(
 
   // TODO handle overflow
   private[this] val _id = new AtomicInteger(BaseStreamId)
-  private[this] def nextId() = _id.getAndAdd(2)
-  private[this] val streams = new ConcurrentHashMap[Int, Netty4ClientStreamTransport]
+  private[this] def nextId(): Int = {
+    val id = _id.getAndAdd(2)
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: Integer overflow kicks in before stream id overflow does:

```
scala> val max = math.pow(2,31)-1
max: Double = 2.147483647E9

scala> val id = new AtomicInteger(max.toInt)
id: java.util.concurrent.atomic.AtomicInteger = 2147483647

scala> id.getAndAdd(2) > max
res10: Boolean = false

scala> id.getAndAdd(2) > max
res11: Boolean = false

scala> id.getAndAdd(2) > max
res12: Boolean = false

scala> id.getAndAdd(2) > max
res13: Boolean = false

scala> id.get
res14: Int = -2147483641
```

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 8]
```diff
@@ -58,59 +73,78 @@ class Netty4ClientDispatcher(
...
    * frames from the transport onto a per-stream receive queue.
    */
-  private[this] val demux = {
-    def loop(): Future[Unit] =
-      if (closed) Future.Unit
-      else transport.read().flatMap { frame =>
-        frame.streamId match {
-          case 0 => log.error(s"dropping ${frame.name} message on the connection")
+  private[this] val reading = {
+    lazy val loop: Http2Frame => Future[Unit] = {
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: why lazy?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 9]
```diff
@@ -0,0 +1,185 @@
+package com.twitter.finagle.buoyant.h2
+package netty4
+
+import com.twitter.finagle.stats.{StatsReceiver, NullStatsReceiver}
+import com.twitter.logging.Logger
+import com.twitter.util.{Future, Promise, Return, Stopwatch, Throw}
+import io.netty.handler.codec.http2._
+
+/**
+ *
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: missing

File: router/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcherTest.scala [hunk 10]
```diff
@@ -13,17 +13,23 @@ import org.scalatest.FunSuite
 import scala.collection.immutable.Queue
 
 class Netty4ClientDispatchTest extends FunSuite with Awaits {
+  // import com.twitter.logging._
```
Comment (adleong) [inline router/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcherTest.scala]: use i.b.t.FunSuite?

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala (hunk 8 above)
Comment (olix0r) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: referenced recursively

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala (hunk 3 above)
Comment (olix0r) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: Yep. There are no IDs left to use on this connection.  Need to start again on a fresh connection.

File: router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala (hunk 2 above)
Comment (olix0r) [inline router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala]: It's not reliable-reliable, but how do you test a negative...  This is really the inverse of `eventually`.  This is my hacky version of `notEventually`

File: router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala [hunk 11]
```diff
@@ -30,26 +34,39 @@ class Netty4ClientDispatcher(
...
   private[this] val writer = Netty4H2Writer(transport)
 
-  // TODO handle overflow
-  private[this] val _id = new AtomicInteger(BaseStreamId)
-  private[this] def nextId() = _id.getAndAdd(2)
-  private[this] val streams = new ConcurrentHashMap[Int, Netty4ClientStreamTransport]
+  private[this] val _id = new AtomicLong(BaseStreamId)
+  private[this] def nextId(): Int = {
+    val id = _id.getAndAdd(2)
+    if (id > MaxStreamId) {
```
Comment (adleong) [inline router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ClientDispatcher.scala]: TIOLI alternative: you could keep the AtomicInt and then check 
```
if (id > MaxStreamId || id < 0)
```

**Comment by olix0r** [comment]
merged https://github.com/BuoyantIO/linkerd/commit/59df8f2fe7a975eedf3dd10d88d43b89614091b9

**Pull request closed without merging**